homeassistant/components/light/zha.py

```diff
@@ -0,0 +1,128 @@
+"""
+Lights on Zigbee Home Automation networks.
+
+For more details on this platform, please refer to the documentation
+at https://home-assistant.io/components/light.zha/
+"""
+import asyncio
+import logging
+
+from homeassistant.components import light, zha
+from homeassistant.util.color import HASS_COLOR_MIN, color_RGB_to_xy
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['zha']
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup Zigbee Home Automation lights."""
+    if discovery_info is None:
+        return
+
+    clusters = discovery_info['clusters']
+    cluster_ids = [c.cluster_id for c in clusters]
+    if 0x0300 in cluster_ids:
+        color = [c for c in clusters if c.cluster_id == 0x0300][0]
+        success, _ = yield from color.read_attributes([0x10], allow_cache=True)
+        if 0x10 in success:
+            discovery_info['num_primaries'] = success[0x10]
+
+    yield from async_add_devices([Light(**discovery_info)])
+
+
+class Light(zha.Entity, light.Light):
+    """ZHA or ZLL light."""
+
+    _domain = light.DOMAIN
+
+    def __init__(self, **kwargs):
+        """Initialize ZHA light."""
+        super().__init__(**kwargs)
+        self._supported_features = 0
+        self._color_temp = None
+        self._xy_color = None
+        self._brightness = None
+        if 8 in self._clusters:
+            self._supported_features |= light.SUPPORT_BRIGHTNESS
+            self._brightness = 0
+        if 0x0300 in self._clusters:
+            # Not sure all color lights necessarily support this directly
+            # Should we emulate it?
+            self._supported_features |= light.SUPPORT_COLOR_TEMP
+            self._color_temp = HASS_COLOR_MIN
+            # Silly heuristic, not sure if it works widely
+            if kwargs.get('num_primaries', 1) >= 3:
+                self._supported_features |= light.SUPPORT_XY_COLOR
+                self._supported_features |= light.SUPPORT_RGB_COLOR
+                self._xy_color = (1.0, 1.0)
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        if self._state == 'unknown':
+            return self._state
+        return bool(self._state)
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Turn the entity on."""
+        duration = 5  # tenths of s
+        if light.ATTR_COLOR_TEMP in kwargs:
+            temperature = kwargs[light.ATTR_COLOR_TEMP]
+            yield from self._clusters[0x0300].command(0xa, temperature,
```

Comment:
I agree that there's too much protocol specific here - mostly because of the hardcoded constants. Would something like this be OK for you?

```python
yield from self._device.color.move_to_color_temp(temperature, duration)
```

Specifically, I'd like to keep the API including the cluster (just referenced in a better way).